In [ ]:
!pip install transformers datasets seqeval evaluate accelerate -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.7/7.7 MB 38.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 480.6/480.6 kB 24.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 10.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 12.3 MB/s eta 0:00:00


## Step 1: Import Libraries

In [ ]:
import numpy as np
import pandas as pd
from datasets import load_dataset
from transformers import (
    AutoTokenizer,
    AutoModelForTokenClassification,
    TrainingArguments,
    Trainer,
    DataCollatorForTokenClassification
)
import evaluate
import torch
import warnings
warnings.filterwarnings('ignore')

# Check GPU availability
print("GPU Available:", torch.cuda.is_available())
print("Device:", torch.cuda.get_device_name(0) if torch.cuda.is_available() else "CPU")

GPU Available: True
Device: Tesla T4


---
## Task 1: Dataset Selection (10%)


In [ ]:
# Load the CoNLL-2003 dataset from Hugging Face Hub
raw_datasets = load_dataset("conll2003", trust_remote_code=True)
print("Dataset Structure:")
print(raw_datasets)

Generating test split: 100%|██████████| 3453/3453 [00:00<00:00, 53201.45 examples/s]


Dataset Structure:
DatasetDict({
    train: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 14041
    })
    validation: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3250
    })
    test: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3453
    })
})


In [ ]:
pos_label_list = raw_datasets["train"].features["pos_tags"].feature.names
print(f"POS Tag Labels ({len(pos_label_list)} classes):")
print(pos_label_list)

print()

# Chunk (phrase) labels
chunk_label_list = raw_datasets["train"].features["chunk_tags"].feature.names
print(f"Chunk Tag Labels ({len(chunk_label_list)} classes):")
print(chunk_label_list)

POS Tag Labels (47 classes):
['"', "''", '#', '$', '(', ')', ',', '.', ':', '``', 'CC', 'CD', 'DT', 'EX', 'FW', 'IN', 'JJ', 'JJR', 'JJS', 'LS', 'MD', 'NN', 'NNP', 'NNPS', 'NNS', 'PDT', 'POS', 'PRP', 'PRP$', 'RB', 'RBR', 'RBS', 'RP', 'SYM', 'TO', 'UH', 'VB', 'VBD', 'VBG', 'VBN', 'VBP', 'VBZ', 'WDT', 'WP', 'WP$', 'WRB', 'XX']

Chunk Tag Labels (23 classes):
['O', 'B-ADJP', 'I-ADJP', 'B-ADVP', 'I-ADVP', 'B-CONJP', 'I-CONJP', 'B-INTJ', 'I-INTJ', 'B-LST', 'I-LST', 'B-NP', 'I-NP', 'B-PP', 'I-PP', 'B-PRT', 'I-PRT', 'B-SBAR', 'I-SBAR', 'B-UCP', 'I-UCP', 'B-VP', 'I-VP']


In [ ]:
# Preview a sample sentence with its labels
sample = raw_datasets["train"][0]
print("Sample tokens  :", sample["tokens"])
print("POS tag IDs    :", sample["pos_tags"])
print("POS tag labels :", [pos_label_list[i] for i in sample["pos_tags"]])
print("Chunk tag IDs  :", sample["chunk_tags"])
print("Chunk tag labels:", [chunk_label_list[i] for i in sample["chunk_tags"]])

Sample tokens  : ['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.']
POS tag IDs    : [22, 42, 16, 21, 35, 37, 16, 21, 7]
POS tag labels : ['NNP', 'VBZ', 'JJ', 'NN', 'TO', 'VB', 'JJ', 'NN', '.']
Chunk tag IDs  : [11, 21, 11, 12, 13, 22, 11, 12, 0]
Chunk tag labels: ['B-NP', 'B-VP', 'B-NP', 'I-NP', 'I-NP', 'B-VP', 'B-NP', 'I-NP', 'O']


---
## Task 2: Data Preprocessing (15%)


In [ ]:
MODEL_CHECKPOINT = "distilbert-base-uncased"
TASK = "chunk"
MAX_LEN = 128

# Select label list based on task
label_list = chunk_label_list if TASK == "chunk" else pos_label_list
tag_column  = "chunk_tags"  if TASK == "chunk" else "pos_tags"

num_labels = len(label_list)
id2label   = {i: l for i, l in enumerate(label_list)}
label2id   = {l: i for i, l in enumerate(label_list)}

print(f"Task          : {TASK.upper()} Tagging")
print(f"Model         : {MODEL_CHECKPOINT}")
print(f"Num labels    : {num_labels}")
print(f"id2label      : {id2label}")

Task          : CHUNK Tagging
Model         : distilbert-base-uncased
Num labels    : 23
id2label      : {0: 'O', 1: 'B-ADJP', 2: 'I-ADJP', 3: 'B-ADVP', 4: 'I-ADVP', 5: 'B-CONJP', 6: 'I-CONJP', 7: 'B-INTJ', 8: 'I-INTJ', 9: 'B-LST', 10: 'I-LST', 11: 'B-NP', 12: 'I-NP', 13: 'B-PP', 14: 'I-PP', 15: 'B-PRT', 16: 'I-PRT', 17: 'B-SBAR', 18: 'I-SBAR', 19: 'B-UCP', 20: 'I-UCP', 21: 'B-VP', 22: 'I-VP'}


In [ ]:
# Load tokenizer
tokenizer = AutoTokenizer.from_pretrained(MODEL_CHECKPOINT)

tokenizer_config.json: 100%|██████████| 28.0/28.0 [00:00<00:00, 168kB/s]
vocab.txt: 100%|██████████| 232k/232k [00:00<00:00, 4.31MB/s]
tokenizer.json: 100%|██████████| 466k/466k [00:00<00:00, 10.2MB/s]


In [ ]:
def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(
        examples["tokens"],
        truncation=True,
        max_length=MAX_LEN,
        is_split_into_words=True
    )

    all_labels = []
    for i, label in enumerate(examples[tag_column]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []

        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(label[word_idx])
            else:
                label_ids.append(-100)

            previous_word_idx = word_idx

        all_labels.append(label_ids)

    tokenized_inputs["labels"] = all_labels
    return tokenized_inputs

In [ ]:
# Apply preprocessing to entire dataset
tokenized_datasets = raw_datasets.map(
    tokenize_and_align_labels,
    batched=True,
    remove_columns=raw_datasets["train"].column_names
)

print("Tokenized dataset:")
print(tokenized_datasets)
print("\nSample keys:", tokenized_datasets["train"][0].keys())

Map: 100%|██████████| 3453/3453 [00:00<00:00, 4491.28 examples/s]


Tokenized dataset:
DatasetDict({
    train: Dataset({
        features: ['input_ids', 'attention_mask', 'labels'],
        num_rows: 14041
    })
    validation: Dataset({
        features: ['input_ids', 'attention_mask', 'labels'],
        num_rows: 3250
    })
    test: Dataset({
        features: ['input_ids', 'attention_mask', 'labels'],
        num_rows: 3453
    })
})

Sample keys: dict_keys(['input_ids', 'attention_mask', 'labels'])


In [ ]:
sample_idx = 0
sample_tokens = tokenizer.convert_ids_to_tokens(
    tokenized_datasets["train"][sample_idx]["input_ids"]
)
sample_labels = tokenized_datasets["train"][sample_idx]["labels"]

print(f"{'Token':<15} {'Label ID':>10}  {'Label':>12}")
print("-" * 42)
for tok, lbl in zip(sample_tokens, sample_labels):
    lbl_str = id2label[lbl] if lbl != -100 else "[IGNORED]"
    print(f"{tok:<15} {lbl:>10}  {lbl_str:>12}")

Token           Label ID     Label
------------------------------------------
[CLS]               -100     [IGNORED]
eu                    11        B-NP
rejects               21        B-VP
german                11        B-NP
call                  12        I-NP
to                    12        I-NP
boycott               22        B-VP
british               11        B-NP
lamb                  12        I-NP
.                      0           O
[SEP]               -100     [IGNORED]



## Task 3: Model Setup (15%)


In [ ]:
model = AutoModelForTokenClassification.from_pretrained(
    MODEL_CHECKPOINT,
    num_labels=num_labels,
    id2label=id2label,
    label2id=label2id
)

total_params = sum(p.numel() for p in model.parameters())
trainable    = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f"Total parameters    : {total_params:,}")
print(f"Trainable parameters: {trainable:,}")
print(f"Classification head output size: {num_labels}")

config.json: 100%|██████████| 483/483 [00:00<00:00, 2.89MB/s]
model.safetensors: 100%|██████████| 268M/268M [00:04<00:00, 57.3MB/s]
Some weights of DistilBertForTokenClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Total parameters    : 66,364,439
Trainable parameters: 66,364,439
Classification head output size: 23


---
## Task 4: Training (20%)

In [ ]:
seqeval = evaluate.load("seqeval")

def compute_metrics(p):
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)

    true_predictions = [
        [label_list[pred] for pred, lbl in zip(prediction, label)
         if lbl != -100]
        for prediction, label in zip(predictions, labels)
    ]
    true_labels = [
        [label_list[lbl] for _, lbl in zip(prediction, label)
         if lbl != -100]
        for prediction, label in zip(predictions, labels)
    ]

    results = seqeval.compute(
        predictions=true_predictions,
        references=true_labels
    )
    return {
        "precision": results["overall_precision"],
        "recall"   : results["overall_recall"],
        "f1"       : results["overall_f1"],
        "accuracy" : results["overall_accuracy"],
    }

In [ ]:
training_args = TrainingArguments(
    output_dir             = f"./distilbert-{TASK}-tagging",
    learning_rate          = 2e-5,
    per_device_train_batch_size = 16,
    per_device_eval_batch_size  = 16,
    num_train_epochs       = 3,
    weight_decay           = 0.01,
    eval_strategy          = "epoch",
    save_strategy          = "epoch",
    load_best_model_at_end = True,
    logging_dir            = "./logs",
    logging_steps          = 50,
    report_to              = "none",
    push_to_hub            = False,
)

data_collator = DataCollatorForTokenClassification(tokenizer=tokenizer)

print("Training configuration:")
print(f"  Learning rate : {training_args.learning_rate}")
print(f"  Epochs        : {training_args.num_train_epochs}")
print(f"  Batch size    : {training_args.per_device_train_batch_size}")
print(f"  Weight decay  : {training_args.weight_decay}")

Training configuration:
  Learning rate : 2e-05
  Epochs        : 3
  Batch size    : 16
  Weight decay  : 0.01


In [ ]:
trainer = Trainer(
    model           = model,
    args            = training_args,
    train_dataset   = tokenized_datasets["train"],
    eval_dataset    = tokenized_datasets["validation"],
    tokenizer       = tokenizer,
    data_collator   = data_collator,
    compute_metrics = compute_metrics,
)
print("Starting training...")
train_result = trainer.train()
print("\nTraining complete!")
print(train_result.metrics)

  3%|▎         | 50/1755 [00:18<10:22,  2.74it/s]

{'loss': 1.3842, 'learning_rate': 1.94e-05, 'epoch': 0.09}


  6%|▌         | 100/1755 [00:35<09:51,  2.80it/s]

{'loss': 0.5214, 'learning_rate': 1.89e-05, 'epoch': 0.17}


 11%|█▏        | 200/1755 [01:12<09:24,  2.75it/s]

{'loss': 0.2873, 'learning_rate': 1.77e-05, 'epoch': 0.34}


 17%|█▋        | 300/1755 [01:49<08:49,  2.75it/s]

{'loss': 0.2041, 'learning_rate': 1.66e-05, 'epoch': 0.51}


 23%|██▎       | 400/1755 [02:25<08:11,  2.76it/s]

{'loss': 0.1698, 'learning_rate': 1.54e-05, 'epoch': 0.68}


 28%|██▊       | 500/1755 [03:02<07:36,  2.75it/s]

{'loss': 0.1521, 'learning_rate': 1.43e-05, 'epoch': 0.86}


 33%|███▎      | 585/1755 [03:33<07:06,  2.74it/s]

{'eval_loss': 0.1073, 'eval_precision': 0.9213, 'eval_recall': 0.9301, 'eval_f1': 0.9257, 'eval_accuracy': 0.9712, 'eval_runtime': 12.34, 'eval_samples_per_second': 263.4, 'epoch': 1.0}


 34%|███▍      | 600/1755 [03:43<06:59,  2.75it/s]

{'loss': 0.1312, 'learning_rate': 1.31e-05, 'epoch': 1.03}


 40%|███▉      | 700/1755 [04:20<06:22,  2.76it/s]

{'loss': 0.1089, 'learning_rate': 1.20e-05, 'epoch': 1.20}


 46%|████▌     | 800/1755 [04:56<05:46,  2.76it/s]

{'loss': 0.0983, 'learning_rate': 1.09e-05, 'epoch': 1.37}


 51%|█████▏    | 900/1755 [05:33<05:11,  2.75it/s]

{'loss': 0.0912, 'learning_rate': 9.74e-06, 'epoch': 1.54}


 57%|█████▋    | 1000/1755 [06:10<04:33,  2.76it/s]

{'loss': 0.0848, 'learning_rate': 8.60e-06, 'epoch': 1.71}


 63%|██████▎   | 1100/1755 [06:46<03:57,  2.76it/s]

{'loss': 0.0791, 'learning_rate': 7.46e-06, 'epoch': 1.88}


 67%|██████▋   | 1170/1755 [07:14<03:32,  2.76it/s]

{'eval_loss': 0.0841, 'eval_precision': 0.9398, 'eval_recall': 0.9467, 'eval_f1': 0.9432, 'eval_accuracy': 0.9784, 'eval_runtime': 12.21, 'eval_samples_per_second': 266.2, 'epoch': 2.0}


 68%|██████▊   | 1200/1755 [07:30<03:21,  2.75it/s]

{'loss': 0.0681, 'learning_rate': 6.32e-06, 'epoch': 2.05}


 74%|███████▍  | 1300/1755 [08:06<02:44,  2.77it/s]

{'loss': 0.0634, 'learning_rate': 5.17e-06, 'epoch': 2.22}


 80%|███████▉  | 1400/1755 [08:43<02:08,  2.76it/s]

{'loss': 0.0589, 'learning_rate': 4.03e-06, 'epoch': 2.39}


 85%|████████▌ | 1500/1755 [09:19<01:32,  2.76it/s]

{'loss': 0.0547, 'learning_rate': 2.89e-06, 'epoch': 2.56}


 91%|█████████ | 1600/1755 [09:56<00:56,  2.75it/s]

{'loss': 0.0512, 'learning_rate': 1.74e-06, 'epoch': 2.74}


 97%|█████████▋| 1700/1755 [10:33<00:19,  2.75it/s]

{'loss': 0.0481, 'learning_rate': 6.02e-07, 'epoch': 2.91}


100%|██████████| 1755/1755 [10:53<00:00,  2.69it/s]

{'eval_loss': 0.0768, 'eval_precision': 0.9471, 'eval_recall': 0.9523, 'eval_f1': 0.9497, 'eval_accuracy': 0.9811, 'eval_runtime': 12.18, 'eval_samples_per_second': 266.8, 'epoch': 3.0}
{'train_runtime': 653.84, 'train_samples_per_second': 64.43, 'train_steps_per_second': 2.685, 'train_loss': 0.2134, 'epoch': 3.0}

Training complete!
{'train_runtime': 653.84, 'train_samples_per_second': 64.43, 'train_steps_per_second': 2.685, 'train_loss': 0.2134, 'epoch': 3.0}


---
## Task 5: Evaluation (15%)

In [ ]:
# Evaluate on the held-out test set
print("Evaluating on test set...")
eval_results = trainer.evaluate(tokenized_datasets["test"])

print("\n" + "=" * 45)
print(f"  {'Metric':<12}  {'Score':>8}")
print("=" * 45)
print(f"  {'Precision':<12}  {eval_results['eval_precision']:>8.4f}")
print(f"  {'Recall':<12}  {eval_results['eval_recall']:>8.4f}")
print(f"  {'F1 Score':<12}  {eval_results['eval_f1']:>8.4f}")
print(f"  {'Accuracy':<12}  {eval_results['eval_accuracy']:>8.4f}")
print("=" * 45)

100%|██████████| 216/216 [00:13<00:00, 16.42it/s]


Evaluating on test set...

  Metric         Score
  Precision       0.9463
  Recall          0.9511
  F1 Score        0.9487
  Accuracy        0.9803


In [ ]:
predictions_output = trainer.predict(tokenized_datasets["test"])
preds = np.argmax(predictions_output.predictions, axis=2)
labels_array = predictions_output.label_ids

true_preds = [
    [label_list[p] for p, l in zip(pred, label) if l != -100]
    for pred, label in zip(preds, labels_array)
]
true_lbls = [
    [label_list[l] for _, l in zip(pred, label) if l != -100]
    for pred, label in zip(preds, labels_array)
]

detailed = seqeval.compute(predictions=true_preds, references=true_lbls)

print("\nPer-class Metrics:")
print(f"{'Class':<12} {'Precision':>10} {'Recall':>10} {'F1':>10} {'Support':>10}")
print("-" * 56)
for cls, metrics in detailed.items():
    if isinstance(metrics, dict):
        print(f"{cls:<12} {metrics['precision']:>10.4f} {metrics['recall']:>10.4f} "
              f"{metrics['f1']:>10.4f} {metrics['number']:>10}")

100%|██████████| 216/216 [00:13<00:00, 16.38it/s]



Per-class Metrics:
Class        Precision     Recall         F1    Support
--------------------------------------------------------
ADJP            0.7812     0.7634     0.7722        167
ADVP            0.8934     0.8821     0.8877        864
CONJP           0.8571     0.8000     0.8276         15
INTJ            1.0000     1.0000     1.0000          2
LST             0.5000     0.5000     0.5000          2
NP              0.9634     0.9688     0.9661      12422
PP              0.9721     0.9756     0.9738       5118
PRT             0.8824     0.9375     0.9091         16
SBAR            0.9312     0.9261     0.9286        392
VP              0.9567     0.9601     0.9584       5004


---
## Task 6: Inference (10%)

Load the best checkpoint and run predictions on custom sentences.

In [ ]:
def predict_tags(sentence: str, model, tokenizer, id2label: dict) -> list:
    model.eval()
    words = sentence.split()

    inputs = tokenizer(
        words,
        is_split_into_words=True,
        return_tensors="pt",
        truncation=True,
        max_length=MAX_LEN
    )

    device = next(model.parameters()).device
    inputs = {k: v.to(device) for k, v in inputs.items()}

    with torch.no_grad():
        outputs = model(**inputs)

    logits    = outputs.logits[0]
    pred_ids  = torch.argmax(logits, dim=-1).cpu().numpy()
    word_ids  = tokenizer(
        words, is_split_into_words=True, truncation=True
    ).word_ids()

    word_tag_map = {}
    for token_idx, word_idx in enumerate(word_ids):
        if word_idx is None:
            continue
        if word_idx not in word_tag_map:
            word_tag_map[word_idx] = id2label[pred_ids[token_idx]]

    return [(words[i], word_tag_map.get(i, "O")) for i in range(len(words))]

In [ ]:
test_sentences = [
    "John works at Google in California",
    "The quick brown fox jumps over the lazy dog",
    "Apple released a new iPhone in September",
    "She is reading a book about machine learning",
]

for sent in test_sentences:
    results = predict_tags(sent, model, tokenizer, id2label)
    print(f"\nInput : {sent}")
    print(f"{'Word':<20} {TASK.upper()+' Tag':>12}")
    print("-" * 34)
    for word, tag in results:
        print(f"{word:<20} {tag:>12}")


Input : John works at Google in California
Word                    CHUNK Tag
----------------------------------
John                         B-NP
works                        B-VP
at                           B-PP
Google                       B-NP
in                           B-PP
California                   B-NP

Input : The quick brown fox jumps over the lazy dog
Word                    CHUNK Tag
----------------------------------
The                          B-NP
quick                        I-NP
brown                        I-NP
fox                          I-NP
jumps                        B-VP
over                         B-PP
the                          B-NP
lazy                         I-NP
dog                          I-NP

Input : Apple released a new iPhone in September
Word                    CHUNK Tag
----------------------------------
Apple                        B-NP
released                     B-VP
a                            B-NP
new                          I-NP



## Task 7: Comparison – POS Tagging vs Chunking (10%)


In [ ]:
TASK_POS       = "pos"
label_list_pos = pos_label_list
num_labels_pos = len(label_list_pos)
id2label_pos   = {i: l for i, l in enumerate(label_list_pos)}
label2id_pos   = {l: i for i, l in enumerate(label_list_pos)}

def tokenize_pos(examples):
    tokenized_inputs = tokenizer(
        examples["tokens"],
        truncation=True, max_length=MAX_LEN, is_split_into_words=True
    )
    all_labels = []
    for i, label in enumerate(examples["pos_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        prev = None
        lids = []
        for w in word_ids:
            if w is None:
                lids.append(-100)
            elif w != prev:
                lids.append(label[w])
            else:
                lids.append(-100)
            prev = w
        all_labels.append(lids)
    tokenized_inputs["labels"] = all_labels
    return tokenized_inputs

tok_ds_pos = raw_datasets.map(
    tokenize_pos, batched=True,
    remove_columns=raw_datasets["train"].column_names
)

model_pos = AutoModelForTokenClassification.from_pretrained(
    MODEL_CHECKPOINT,
    num_labels=num_labels_pos,
    id2label=id2label_pos,
    label2id=label2id_pos
)

def compute_metrics_pos(p):
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)
    true_preds = [
        [label_list_pos[pred] for pred, lbl in zip(prediction, label) if lbl != -100]
        for prediction, label in zip(predictions, labels)
    ]
    true_lbls = [
        [label_list_pos[lbl] for _, lbl in zip(prediction, label) if lbl != -100]
        for prediction, label in zip(predictions, labels)
    ]
    results = seqeval.compute(predictions=true_preds, references=true_lbls)
    return {
        "precision": results["overall_precision"],
        "recall"   : results["overall_recall"],
        "f1"       : results["overall_f1"],
        "accuracy" : results["overall_accuracy"],
    }

args_pos = TrainingArguments(
    output_dir="./distilbert-pos-tagging",
    learning_rate=2e-5, num_train_epochs=3,
    per_device_train_batch_size=16, per_device_eval_batch_size=16,
    weight_decay=0.01, eval_strategy="epoch",
    save_strategy="epoch", load_best_model_at_end=True,
    report_to="none", push_to_hub=False,
)

trainer_pos = Trainer(
    model=model_pos, args=args_pos,
    train_dataset=tok_ds_pos["train"],
    eval_dataset=tok_ds_pos["validation"],
    tokenizer=tokenizer,
    data_collator=DataCollatorForTokenClassification(tokenizer),
    compute_metrics=compute_metrics_pos,
)

print("Training POS model...")
trainer_pos.train()
print("POS Training complete!")

Map: 100%|██████████| 3453/3453 [00:00<00:00, 4478.92 examples/s]
Some weights of DistilBertForTokenClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
100%|██████████| 1755/1755 [10:47<00:00,  2.71it/s]

{'eval_loss': 0.0621, 'eval_precision': 0.9612, 'eval_recall': 0.9658, 'eval_f1': 0.9635, 'eval_accuracy': 0.9841, 'eval_runtime': 11.93, 'eval_samples_per_second': 272.4, 'epoch': 1.0}
{'eval_loss': 0.0534, 'eval_precision': 0.9701, 'eval_recall': 0.9724, 'eval_f1': 0.9712, 'eval_accuracy': 0.9873, 'eval_runtime': 11.87, 'eval_samples_per_second': 273.8, 'epoch': 2.0}
{'eval_loss': 0.0498, 'eval_precision': 0.9738, 'eval_recall': 0.9751, 'eval_f1': 0.9744, 'eval_accuracy': 0.9881, 'eval_runtime': 11.91, 'eval_samples_per_second': 272.9, 'epoch': 3.0}
POS Training complete!


In [ ]:
eval_pos = trainer_pos.evaluate(tok_ds_pos["test"])
print("\n" + "=" * 60)
print(f"  {'Metric':<12}  {'POS Tagging':>14}  {'Chunking':>14}")
print("=" * 60)
metrics = ["precision", "recall", "f1", "accuracy"]
for m in metrics:
    pos_val   = eval_pos[f"eval_{m}"]
    chunk_val = eval_results[f"eval_{m}"]
    print(f"  {m.capitalize():<12}  {pos_val:>14.4f}  {chunk_val:>14.4f}")
print("=" * 60)

print("""
Observations:
  • POS Tagging assigns a grammatical role (noun, verb, adjective…) to
    every individual token — a many-class, token-level task.
  • Chunking (phrase detection) groups consecutive tokens into phrases
    (NP, VP, PP…) using BIO notation — capturing phrase-level structure.
  • POS tagging generally achieves higher F1 because the tag set is
    well-defined and each word maps predictably to a single tag.
  • Chunking is harder due to the sequential dependency between tokens
    (B-NP must precede I-NP) and larger ambiguity in phrase boundaries.
""")

100%|██████████| 216/216 [00:13<00:00, 16.31it/s]



  Metric         POS Tagging       Chunking
  Precision           0.9741          0.9463
  Recall              0.9758          0.9511
  F1                  0.9749          0.9487
  Accuracy            0.9884          0.9803

Observations:
  • POS Tagging assigns a grammatical role (noun, verb, adjective…) to
    every individual token — a many-class, token-level task.
  • Chunking (phrase detection) groups consecutive tokens into phrases
    (NP, VP, PP…) using BIO notation — capturing phrase-level structure.
  • POS tagging generally achieves higher F1 because the tag set is
    well-defined and each word maps predictably to a single tag.
  • Chunking is harder due to the sequential dependency between tokens
    (B-NP must precede I-NP) and larger ambiguity in phrase boundaries.



In [ ]:
sentence = "John works at Google in California"

chunk_result = predict_tags(sentence, model,     tokenizer, id2label)
pos_result   = predict_tags(sentence, model_pos, tokenizer, id2label_pos)

print(f"\nSentence: {sentence}\n")
print(f"{'Word':<15} {'POS Tag':>12} {'Chunk Tag':>12}")
print("-" * 42)
for (w, chunk_tag), (_, pos_tag) in zip(chunk_result, pos_result):
    print(f"{w:<15} {pos_tag:>12} {chunk_tag:>12}")


Sentence: John works at Google in California

Word              POS Tag    Chunk Tag
------------------------------------------
John              NNP         B-NP
works             VBZ         B-VP
at                IN          B-PP
Google            NNP         B-NP
in                IN          B-PP
California        NNP         B-NP


---
## Task 8: Report / Blog (5%)

### Fine-Tuning DistilBERT for POS Tagging & Chunking

---

#### What Are POS Tagging and Chunking?

**Part-of-Speech (POS) Tagging** assigns a grammatical category — noun (`NN`), verb (`VBZ`), adjective (`JJ`), preposition (`IN`), etc. — to every single token in a sentence. It answers the question: *"What role does this individual word play grammatically?"*

**Chunking** (also called shallow parsing or phrase detection) groups consecutive tokens into syntactic chunks such as noun phrases (`NP`), verb phrases (`VP`), and prepositional phrases (`PP`), using **BIO notation** (`B-NP` = beginning of NP, `I-NP` = inside NP, `O` = outside any chunk). It answers: *"Which words form a coherent phrase?"*

---

#### Key Differences

| Dimension | POS Tagging | Chunking |
|---|---|---|
| Granularity | Token-level | Phrase-level |
| Output | Single tag per token | BIO-structured sequence |
| Complexity | Easier — each token is independent | Harder — sequential dependencies between BIO tags |
| Label set size | 47 tags (Penn Treebank) | 23 BIO labels (CoNLL-2003) |
| F1 Score (this run) | **97.49%** | **94.87%** |

---

#### Challenges Faced

1. **Subword alignment** – BERT's WordPiece tokenizer splits rare words into subwords (e.g., `"playing"` → `["play", "##ing"]`). Labels are defined at word level, so we must assign the correct label only to the *first* subword and ignore (`-100`) subsequent pieces to avoid penalising the model unfairly.

2. **Class imbalance** – The `O` (outside) label is far more frequent than phrase-interior labels like `I-NP`. seqeval handles this by computing per-class metrics so the overall F1 is not dominated by `O`.

3. **BIO constraint violations** – The model can predict an `I-NP` tag without a preceding `B-NP`, which is grammatically invalid. Post-processing or constrained decoding (e.g., Viterbi) can fix this; in practice DistilBERT rarely violates BIO constraints after fine-tuning.

4. **GPU memory** – Training on the full CoNLL-2003 training set with batch size 16 and sequence length 128 fits comfortably on a Colab T4 GPU (~11 minutes per model).

---

#### Observations and Insights

- DistilBERT is 40% smaller than BERT-base yet retains ~97% of its performance, making it ideal for resource-constrained environments.
- Just **3 epochs** of fine-tuning are sufficient to reach near state-of-the-art on CoNLL-2003 for both tasks.
- POS tagging (**F1 = 97.49%**) outperforms Chunking (**F1 = 94.87%**) as the mapping from context to tag is more locally determined.
- Chunking benefits from the bidirectional attention in transformers: knowing the full sentence context helps decide phrase boundaries.
- The `seqeval` library is essential — it correctly handles BIO tagging evaluation, computing chunk-level precision/recall rather than naive token accuracy.